### DISCUSSION TITLE
relax.op.dense 的定义在哪里？

### DISCUSSION THREAD

**Discussion by DzAvril** [mlc-ai/mlc-zh#93] [Q&A] [answered]
第五课最后的代码
```python
def map_nn_linear_op(bb, node_map, node, nn_mod):
    x = node_map[node.args[0]]
    w = map_param(nn_mod.weight)
    if nn_mod.bias is not None:
        b = map_param(nn_mod.bias)
    y = bb.emit(relax.op.dense(x, w))
    return bb.emit(relax.op.add(y, b))
```
请问这里的relax.op.dense是在哪里定义的？我在relax仓库里没有找到。这里使用了te.op替代了TE，是不是te.op是用tvm script定义的算子？

**Comment by Hzfengsy** [comment, marked as answer]
https://github.com/mlc-ai/relax/blob/495ac60100efda81a1d507d228a9ada5b9ed7392/python/tvm/relax/op/nn/nn.py#L22

**Reply by DzAvril** [reply]
感谢回答，能否再指点yixian "relax.nn.dense"对应的compute在哪里定义的？

**Reply by sisiwuxi** [reply]
def map_nn_relu_op(bb, node_map, node, nn_mod):
      A = node_map[node.args[0]]
      return bb.emit(relax.op.nn.relu(A))

  def map_nn_linear_op(bb, node_map, node, nn_mod):
      x = node_map[node.args[0]]
      w = map_param(nn_mod.weight)
      if nn_mod.bias is not None:
          b = map_param(nn_mod.bias)
      y = bb.emit(relax.op.nn.dense(x, w))
      return bb.emit(relax.op.add(y, b))

**Reply by Hzfengsy** [reply]
> 感谢回答，能否再指点yixian "relax.nn.dense"对应的compute在哪里定义的？

很抱歉没有及时回复。目前还并没有对op做正式的支持（包含compute, lowering, VM等)。本课程中我们想表达的是，我们的Torch integrator有能力转换成high-level op，在未来可能有用
